cloudformation/functions/get_groups_from_policy.py

```diff
@@ -0,0 +1,98 @@
+import sys
+
+from json import loads
+from json.decoder import JSONDecodeError
+
+
+INVALID_OPERATORS = (
+    "StringNotEquals", "ForAnyValue:StringNotEquals",
+    "StringNotLike", "ForAnyValue:StringNotLike"
+)
+UNGLOBBABLE_OPERATORS = ("StringEquals", "ForAnyValue:StringEquals")
+VALID_OPERATORS = (
+    "StringEquals", "ForAnyValue:StringEquals",
+    "StringLike", "ForAnyValue:StringLike",
+)
+
+VALID_AMRS = (
+    "auth-dev.mozilla.auth0.com/:amr",
+    "auth.mozilla.auth0.com/:amr",
+)
+VALID_FEDERATED_PRINCIPAL_KEYS = (
+    "arn:aws:iam::656532927350:oidc-provider/auth-dev.mozilla.auth0.com/",
+    "arn:aws:iam::371522382791:oidc-provider/auth.mozilla.auth0.com/",
+)
+
+
+# Python 2 compatibility
+if sys.version_info > (3, 0):
+    basestring = str
+
+
+def get_groups_from_policy(policy) -> list:
+    # groups will be stored as a set to prevent duplicates and then return
+    # a list when everything is finished
+    policy_groups = set()
+
+    # be flexible on being passed a dictionary (parsed policy) or a string
+    # (unparsed policy)
+    if isinstance(policy, basestring):
+        try:
+            policy = loads(policy)
+        except JSONDecodeError:
+            return []
+
+    if not isinstance(policy, dict):
+        raise ValueError
+
+    # If policy lacks a statement, we can bail out
+    if 'Statement' not in policy:
+        return []
+
+    for statement in policy["Statement"]:
+        if (statement.get("Effect") != "Allow" or
+                statement.get("Action") != "sts:AssumeRoleWithWebIdentity" or
+                statement.get('Principal', {}).get('Federated') not in
+                VALID_FEDERATED_PRINCIPAL_KEYS):
+            continue
+
+        operator_count = 0
+        for operator in statement.get("Condition", {}).keys():
+            # StringNotLike, etc. are not supported
+            if operator in INVALID_OPERATORS:
+                return []
+            # Is a valid operator and contains a valid :amr entry
+            elif (operator in VALID_OPERATORS and
+                    any(amr in VALID_AMRS for amr in
+                        statement["Condition"][operator].keys())):
+                operator_count += 1
+
+        # Multiple operators are not supported
+        if operator_count != 1:
+            return []
+
+        # For clarity:
+        # operator --> StringEquals, ForAnyValue:StringLike
+        # conditions --> dictionary mapping, e.g. StringEquals: {}
+        # condition: auth-dev.mozilla.auth0.com/:amr
+        operator_count = 0
```

Comment:
I'm not sure I understand this?

Reply:
I don't think this assignment is necessary is it since it only need be done once in the loop and that happens on line 59? I may not have understood why it's done a second time

Reply:
I just did a significant rewrite, but I'm still very confused. Maybe we could have a conversation about it?

Reply:
for what its worth its the only point that also confuses me (ie, why its reset to zero here)

Reply:
A policy can have any number of statements, and each statement can have one or more operators. We can only work with single operators (eg no `AND`), and it's resetting to zero with each statement.